antismash/modules/nrps_pks/transat_ks_analysis/transat_ks_analysis.py

```diff
@@ -0,0 +1,261 @@
+    return(leaf2clade, clade2ann)
+
+
+def choose_elder(parent: Any, grandparent: Any) -> Any:
+    """ Selects either parent or grandparent for monophyly test """
+    ## Find the proper clade elder, if more than two siblings use parent, otherwise use grandparent
+    if parent is None:
+        return None
+    if len(parent.get_terminals()) > 2:
+        return parent
+    if grandparent is None:
+        return None
+    return grandparent
+
+
+def get_transpact_clade(query_name: str, tree: Any, funclades: Dict[str, str]) -> str:
+    """
+    tree: Bio.Phylo.Newick.Tree
+    """
+    ## Remove placements that aren't the query in question and note query's elders
+    newtree = copy.deepcopy(tree) ## necessary to not screw up original tree for subsequent queries
+    parent, grandparent = None, None
+    for leaf in newtree.get_terminals():
+        if leaf.name == query_name:
+            node_path = newtree.get_path(leaf)
+            parent, grandparent = node_path[-2], node_path[-3]
+        else:
+            ln = leaf.name.split("_")
+            if re.match("^#\d+$", ln[-2]) is not None:
+                newtree.prune(leaf)
+    elder = choose_elder(parent, grandparent)
+    if elder is None:
+        raise ValueError("No leaf named %s in tree terminals." % query_name)
+    ## Count number of occurances of each clade in elder descendants
+    clade_count = {}  # type: Dict[str, int]
+    for leaf in elder.get_terminals():
+        if leaf.name != query_name:
+            if funclades[leaf.name] in clade_count:
+                clade_count[funclades[leaf.name]] += 1
+            else:
+                clade_count[funclades[leaf.name]] = 1
+    clade_assignment = 'clade_not_conserved'
+    if len(clade_count) == 1: ## clade consensus, monophyly
+        clade_assignment = list(clade_count)[0]
+    return clade_assignment
+
+
+def get_best_transpact(totalmass: Dict[str, float],
+                       masscutoff: float) -> Tuple[str, float]:
+    """ Returns clade, spec and score for a transPACT query """
+    best_clade: str = 'None'
+    best_mass: float = float(0)
+    for cld in totalmass:
+        if totalmass[cld] > best_mass:
+            best_clade, best_mass = cld, totalmass[cld]
+    clade: str = 'clade_not_conserved'
+    score: float = 0.0
+    if best_clade != 'clade_not_conserved' and best_mass >= masscutoff:
+        clade = best_clade
+        score = round(best_mass, 2)
+    return clade, score
+
+
+def get_totalmass(tree: Any,
+                  tree_hits: Dict[str, Any],
+                  funclades: Dict[str, str]) -> Dict[str, float]:
+    """ Returns the totalmass dict for each clade """
+    totalmass: Dict[str, float] = {}
+    query_prefix = None
+    for placement_num in tree_hits:
+        if query_prefix is None:
+            query_prefix = re.sub(r"^(.+)_#\d+_M=\d+?\.?\d*$", "\g<1>", tree_hits[placement_num].name)
+        mass = float(re.sub(r"^.+#\d+_M=(\d+?\.?\d*)$", "\g<1>", tree_hits[placement_num].name))
```

Comment:
Rather than a regex, why not just use `mass = float(tree.hits[placement_num].name.rsplit("M=", 1)[1])`?